src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,135 @@
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final Object[] EMPTY_ELEMENT_DATA = {};
+    private Object[] elementList;
+    private int size;
+
+    public ArrayList() {
+        this.elementList = EMPTY_ELEMENT_DATA;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementList.length) {
+            elementList = grow();
+        }
+        elementList[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can not add element, "
+                    + "index is out of bounds.");
+        }
+        if (size == elementList.length) {
+            elementList = grow();
+        }
+        System.arraycopy(elementList, index, elementList, index + 1, size - index);
+        elementList[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] listAsArray = list.toArray();
+        if (listAsArray.length > elementList.length - size) {
+            elementList = grow(listAsArray.length + size);
+        }
+        System.arraycopy(listAsArray, 0, elementList, size, listAsArray.length);
+        size += listAsArray.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can not get element, "
+                    + "index is out of bounds.");
+        }
+        return (T) elementList[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can not set element, "
+                    + "index is out of bounds.");
+        }
+        elementList[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can not remove element, "
+                    + "index is out of bounds.");
+        }
+        int newSize = size - 1;
+        final Object oldValue = elementList[index];
+        if (newSize > index) {
+            System.arraycopy(elementList, index + 1, elementList, index, newSize - index);
+        }
+        size = newSize;
+        elementList[size] = null;
+        return (T) oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if ((element == elementList[i])
+                    || (element != null && element.equals(elementList[i]))) {
+                index = i;
+                break;
+            }
+        }
+        if (index == -1) {
+            throw new NoSuchElementException("Element not found");
+        }
```

Comment:
throw exception unconditionally here
Returning early is usually a [preferred approach](https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement) in SE